### PR TITLE
WIP: System.Security: Optimize key imports for Common Crypto and OpenSSL.

### PR DESCRIPTION
**Pull request by filipnavara**
Includes #31627, just trying to run CI, will rebase later. Only the last three commits are relevant to this PR.

### CODE REVIEW

File: src/Common/src/Internal/Cryptography/AsymmetricAlgorithmHelpers.Der.cs [hunk 1]
```diff
@@ -63,30 +65,22 @@ public static int BitsToBytes(int bitLength)
             return byteLength;
         }
 
-        private static void CopySignatureField(byte[] signatureField, byte[] response, int offset, int fieldLength)
+        private static void CopySignatureField(ReadOnlySpan<byte> signatureField, byte[] response, int offset, int fieldLength)
```
Comment (krwq) [inline src/Common/src/Internal/Cryptography/AsymmetricAlgorithmHelpers.Der.cs]: I'd make second arg a Span

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/ECDsaX509SignatureGenerator.cs [hunk 2]
```diff
@@ -104,7 +113,7 @@ protected override PublicKey BuildPublicKey()
             
             return new PublicKey(
                 ecPublicKey,
-                new AsnEncodedData(ecPublicKey, DerEncoder.EncodeOid(curveOid)),
+                new AsnEncodedData(ecPublicKey, curveOidEncoded),
```
Comment (krwq) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/ECDsaX509SignatureGenerator.cs]: not seeing any gain from this change - previous code was shorter though

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/RSAPkcs1X509SignatureGenerator.cs [hunk 3]
```diff
@@ -47,7 +42,7 @@ internal static PublicKey BuildPublicKey(RSA rsa)
                 // This is due to one version of the ASN.1 not including OPTIONAL, and that was
                 // the version that got predominately implemented for RSA. Now it's convention.
                 new AsnEncodedData(oid, new byte[] { 0x05, 0x00 }),
-                new AsnEncodedData(oid, rsaPublicKey));
+                new AsnEncodedData(oid, rsa.ExportRSAPublicKey()));
```
Comment (krwq) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/RSAPkcs1X509SignatureGenerator.cs]: is this guaranteed to be DER?

**Comment by filipnavara** [comment]
@krwq Your comments are kinda valid, but they are on the commit that belong to different PR (#31602). I know it's a mess, but my patch queue was getting too long so I wanted to run it through CI before further reviews. The linked PR doesn't attempt to be performance optimization, just code sharing and removing remaining `DerEncoder`/`DerSequenceReader` usages.

File: src/Common/src/Internal/Cryptography/AsymmetricAlgorithmHelpers.Der.cs (hunk 1 above)
Comment (filipnavara) [reply src/Common/src/Internal/Cryptography/AsymmetricAlgorithmHelpers.Der.cs]: Makes sense.

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/ECDsaX509SignatureGenerator.cs (hunk 2 above)
Comment (filipnavara) [reply src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/ECDsaX509SignatureGenerator.cs]: Sorry, that is part of the squashed commits from different PR (#31602). That PR is part of series of patches that remove DerEncoder and DerSequenceReader in favor of AsnReader / AsnWriter / AsnSerializer.

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/RSAPkcs1X509SignatureGenerator.cs (hunk 3 above)
Comment (filipnavara) [reply src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/RSAPkcs1X509SignatureGenerator.cs]: I don't think it's guaranteed to be DER, but it should be guaranteed to be valid ASN.1 BER representation that belongs in that place (RFC 8017).

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/ECDsaX509SignatureGenerator.cs (hunk 2 above)
Comment (krwq) [reply src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/ECDsaX509SignatureGenerator.cs]: fair enough, this specific case would make sense to replace internals of DerEncoder.EncodeOid to use AsnWriter

**Comment by filipnavara** [comment]
Will resubmit after PR #31579 and #31602 are merged.

**Pull request closed without merging**